vertical-pod-autoscaler/pkg/recommender/routines/recommender_routines_test.go

```diff
@@ -0,0 +1,267 @@
+		resourceEstimation = GetPercentileResourceEstimation(cpuPercentile, memPercentile,
+			cpuHistogram, memoryPeaksHistogram)
+
+		if resourceEstimation[model.ResourceMemory] >= model.ResourceAmount(memorySample) {
+			break
+		}
+		surgeSamplesCount++
+	}
+
+	return resourceEstimation, surgeSamplesCount
+}
+
+func GetPercentileResourceEstimation(cpuPercentile float64, memPercentile float64,
+	cpuHistogram util.Histogram, memoryHistogram util.Histogram) model.Resources {
+
+	estimator := logic.NewPercentileEstimator(cpuPercentile, memPercentile)
+
+	return estimator.GetResourceEstimation(
+		&model.AggregateContainerState{
+			AggregateCPUUsage:    cpuHistogram,
+			AggregateMemoryPeaks: memoryHistogram,
+		})
+}
+
+func getTestVPAObj(cluster *model.ClusterState, container model.ContainerID) *model.Vpa {
+	vpa := test.VerticalPodAutoscaler().WithName(testVpaID.VpaName).
+		WithNamespace(testVpaID.Namespace).WithContainer(container.ContainerName).WithUpdateMode(testUpdateModeAuto).Get()
+
+	labelSelector, _ := metav1.ParseToLabelSelector(testSelectorStr)
+	parsedSelector, _ := metav1.LabelSelectorAsSelector(labelSelector)
+	err := cluster.AddOrUpdateVpa(vpa, parsedSelector)
+	if err != nil {
+		klog.Fatalf("AddOrUpdateVpa() failed: %v", err)
+	}
+
+	return cluster.Vpas[testVpaID]
+}
+
+// Verifies that recommendation does not occur and remains Nil
+// if only memory samples are added onto the aggregate container state
+// histogram. The variable TotalSamplesCount is counted only for CPU
+// samples and a similar for memory should be introduced.
+func TestRecommendationForZeroTotalSamplesCount(t *testing.T) {
+	cluster := model.NewClusterState()
+	cluster.AddOrUpdatePod(testPodID1, testLabels, apiv1.PodRunning)
+	container := model.ContainerID{PodID: testPodID1, ContainerName: "app-A"}
+
+	assert.NoError(t, cluster.AddOrUpdateContainer(container, testRequest))
+	assert.NoError(t, addTestMemorySample(cluster, container, 2e9, testTimestamp))
+
+	pod := cluster.Pods[testPodID1]
+	aggregateStateKey := cluster.MakeAggregateStateKey(pod, "app-A")
+	aggregateResources := cluster.GetAggregateStateMap()
+	// Note: This returns TotalSamplesCount as 0 as we are only adding memory sample
+	// This is handled in another test case (TestTotalSamplesCountMemorySamplesOnly).
+	// Ignorning and continuing to seek recommendations for now. Fix it to equal to 1.
+	assert.Equal(t, 0, aggregateResources[aggregateStateKey].TotalSamplesCount)
+
+	recommender := logic.CreatePodResourceRecommender()
+	vpaObj := getTestVPAObj(cluster, container)
+	recommendedResources := recommender.GetRecommendedPodResources(GetContainerNameToAggregateStateMap(vpaObj))
+	assert.NotEmpty(t, recommendedResources)
+
+}
+
+// Verifies the number of samples it takes for the current
+// VPA recommender estimation to beat the current surge in usage.
+// The surge is determined and samples w.r.t that is added using
+// model.OOMBumpUpRatio
+func TestAddMemorySurgeWithOOMBumpUpRatio(t *testing.T) {
```

Comment:
These test case functions follow a pattern. Can you have a helper function to capture and reuse that pattern (along the lines of [template method pattern](https://www.geeksforgeeks.org/template-method-design-pattern/))?

For example,
```go
func doTestRecommender(
        initialCPUSamples, initialMemorySamples resourceSamples,
        assertInitialRecommendations func(RecommendedPodResources),
        doTestAction func(),
        assertPostTestActionRecommendations func(RecommendedPodResources),
) {
       ...
       addTestCPUHistogram(cpuHistogram, initialCPUSamples)
       addTestMemoryHistogram(memoryPeakHistogram, initialCPUSamples)
       ...
       assertInitialRecommendations(recommender.GetRecommendedPodResources(GetContainerNameToAggregateStateMap(vpaObj)))

       doTestActionFn()
       
       assertPostTestActionRecommendations(recommender.GetRecommendedPodResources(GetContainerNameToAggregateStateMap(vpaObj)))
}
```

It can then be called like below.
```go
func TestAddMemorySurgeWithOOMBumpUpRatio(t *testing.T) {
        doTestRecommender(
                []resourceSamples{
                        {1.0, 1.0, timestamp1},
                        {0.9, 1.0, timestamp2},
                },
                []resourceSamples{
                        {1.0, 1.0, timestamp1},
                        {0.9, 1.0, timestamp2},
                },
                func(res RecommendedPodResources) { /* do asserts on initial recommendations */ },
                func() {
                        //This could be direct manipulation of the histograms of indirect manipulation by calling other functions that manipulate the histogram like RecordOOM
                        addTestMemoryHistogram(memoryPeakHistogram, []resourceSamples{
                                 { 1.0, 1.0, timestamp3 },
                        })
                },
                func(res RecommendedPodResources) { /* do asserts on post test action recommendations */ },
        )
}
```

Reply:
ACK.
Converted the tests to use Template method pattern as suggested.
Thanks for the reference example!